src/main/java/games/strategy/net/nio/Encoder.java

```diff
@@ -90,4 +91,17 @@ private void write(final MessageHeader header, final ObjectOutputStream out, fin
     }
     out.reset();
   }
+
+  /**
+   * Provide access to the raw buffer.
+   */
+  private static final class ByteArrayOutputStream2 extends ByteArrayOutputStream {
```

Comment:
Yeah, if you look at how the `byte[]` is used in the `SocketWriteData` constructor, it's simply copied to a `ByteBuffer`.  So that would imply it doesn't really matter if `SocketWriteData` receives a direct reference to the underlying `byte[]` or a copy.  It was probably just done for some kind of perceived performance improvement.

I will look into removing `ByteArrayOutputStream2` in a separate PR.